internal/imageprovider/imageprovider.go

```diff
@@ -1,95 +1,120 @@
 }
 
-// initCache initializes the bird image cache with the given image provider
-func initCache(e ImageProvider) *BirdImageCache {
+// SetNonBirdImageProvider allows setting a custom ImageProvider for non-bird entries
+func (c *BirdImageCache) SetNonBirdImageProvider(provider ImageProvider) {
+	c.nonBirdImageProvider = provider
+}
+
+// SetImageProvider allows setting a custom ImageProvider for testing purposes.
+func (c *BirdImageCache) SetImageProvider(provider ImageProvider) {
+	c.birdImageProvider = provider
+}
+
+// initCache initializes a new BirdImageCache with the given ImageProvider.
+func InitCache(e ImageProvider, t *telemetry.Metrics) *BirdImageCache {
 	return &BirdImageCache{
 		birdImageProvider:    e,
-		nonBirdImageProvider: &emptyImageProvider{}, // TODO: Use a real image provider for non-birds
+		nonBirdImageProvider: &emptyImageProvider{},
+		metrics:              t.ImageProvider,
 	}
 }
 
-// CreateDefaultCache creates a new bird image cache with the default image provider
-func CreateDefaultCache() (*BirdImageCache, error) {
+// CreateDefaultCache creates a new BirdImageCache with the default WikiMedia image provider.
+func CreateDefaultCache(metrics *telemetry.Metrics) (*BirdImageCache, error) {
 	provider, err := NewWikiMediaProvider()
 	if err != nil {
 		return nil, err
 	}
-	return initCache(provider), nil
+	return InitCache(provider, metrics), nil
 }
 
-// Get retrieves the bird image for a given scientific name from the cache or fetches it if not present
+// Get retrieves the BirdImage for a given scientific name from the cache,
+// fetching it from the provider if not already cached.
 func (c *BirdImageCache) Get(scientificName string) (BirdImage, error) {
-	// Check if the bird image is already in the cache
+	if c.metrics == nil {
+		log.Println("Warning: ImageProviderMetrics is nil")
+	}
+
 	if birdImage, ok := c.dataMap.Load(scientificName); ok {
+		if c.metrics != nil {
+			c.metrics.IncrementCacheHits()
+			//log.Printf("Cache hit for %s", scientificName)
+		}
 		return birdImage.(BirdImage), nil
 	}
 
-	// Use a per-item mutex to ensure only one query is performed per item
+	if c.metrics != nil {
+		c.metrics.IncrementCacheMisses()
+		//log.Printf("Cache miss for %s", scientificName)
+	}
+
 	mu, _ := c.dataMutexMap.LoadOrStore(scientificName, &sync.Mutex{})
 	mutex := mu.(*sync.Mutex)
 
 	mutex.Lock()
 	defer mutex.Unlock()
 
-	// Check again if bird image is cached after acquiring the lock
 	if birdImage, ok := c.dataMap.Load(scientificName); ok {
+		c.metrics.IncrementCacheHits()
 		return birdImage.(BirdImage), nil
 	}
 
-	// Fetch the bird image from the image provider
 	fetchedBirdImage, err := c.fetch(scientificName)
 	if err != nil {
-		// TODO for now store an empty result in the cache to avoid future queries that would fail.
-		// In the future, look at the error and decide if it was caused by networking and is recoverable.
-		// And if it was, do not store the empty result in the cache.
```

Comment:
Still think this todo is valid and should not be removed

Reply:
Some comments were removed unintentionally, like this

Reply:
Ah ok I understand :)